### DISCUSSION TITLE
eModbus for Arduino

### DISCUSSION THREAD

**Discussion by arcanine84** [eModbus/eModbus#316] [Q&A] [answered]
I'm trying to implement the examples, but I only have one ESP32, and the other is an UNO.  When I try to compile for the UNO, there is an error, because Arduino does not support the vector construct.

```
In file included from c:\Users\Jeff\Documents\Arduino\libraries\eModbus\src\CoilData.cpp:6:0:
c:\Users\Jeff\Documents\Arduino\libraries\eModbus\src\CoilData.h:9:10: fatal error: vector: No such file or directory
 #include <vector>
          ^~~~~~~~
compilation terminated.

exit status 1

Compilation error: exit status 1
```

I am extremely new to all of this, so I think my options are:
1. Purchase another ESP32. (I'd rather not, plus I want to use a TFT shield that I have for the UNO.)
2. Find a library that implements vectors for Arduino. (I was thinking [this one](https://github.com/maniacbug/StandardCplusplus/tree/master).)
3. Implement my own vector.  (Too advanced for my current abilities and understanding...just being real.)

Any thoughts would be appreciated.

**Comment by Miq1** [comment, marked as answer]
As you found out the hard way, the regular Arduino core does not support the C++ STL in all its glory. There are implementations of the STL, but although these are a nice piece of work, I would not consider them fit to a regular application. The memory of the classic Arduinos is too limited to leave enough room for relevant code when a STL is incorporated. This was a reason to implement eModbus for the Espressif MCUs only that are providing much more memory and an almost complete STL in the core.

Long story short - there are Modbus RTU libraries suited to be used on Arduinos, so you better take a look for those. As the Modbus RTU protocol is standardized, these will work nicely in combination with the ESP32 running eModbus.
